src/mjolnir/restrictionbuilder.cc

```diff
@@ -23,220 +23,155 @@ using namespace valhalla::midgard;
-          graphids.push_back(g_id);
-
-          currentNode = de->endnode();
-          // get the new tile if needed.
-          if (endnodetile->id() != currentNode.Tile_Base()) {
-            lock.lock();
-            endnodetile = reader.GetGraphTile(currentNode);
-            lock.unlock();
-          }
-
-          // get new end node and start over.
-          n_info = endnodetile->node(currentNode);
-          break;
-        } else if (begin_wayid == current_offset.wayid()) {
-          if (visited_set.find(currentNode) == visited_set.end()) {
-            prev_Node = currentNode;
-            visited_set.emplace(currentNode);
-            graphids.push_back(g_id);
-
-            currentNode = de->endnode();
-            // get the new tile if needed.
-            if (endnodetile->id() != currentNode.Tile_Base()) {
-              lock.lock();
-              endnodetile = reader.GetGraphTile(currentNode);
-              lock.unlock();
-            }
-
-            // get new end node and start over.
-            n_info = endnodetile->node(currentNode);
-            j = 0;
-            bBeginFound = true;
-            continue;
-          }
-        }
+namespace {
+
+struct EdgeId {
+  uint64_t way_id;
+  GraphId graph_id;
+};
+
+bool ExpandFromNode(GraphReader& reader,
+                    std::mutex& lock,
+                    GraphId& last_node,
+                    std::vector<EdgeId>& edge_ids,
+                    const std::vector<uint64_t>& way_ids,
+                    size_t way_id_index,
+                    const graph_tile_ptr& prev_tile,
+                    GraphId prev_node,
+                    GraphId current_node);
+
+bool ExpandFromNodeInner(GraphReader& reader,
+                         std::mutex& lock,
+                         GraphId& last_node,
+                         std::vector<EdgeId>& edge_ids,
+                         const std::vector<uint64_t>& way_ids,
+                         size_t way_id_index,
+                         const graph_tile_ptr& tile,
+                         GraphId prev_node,
+                         GraphId current_node,
+                         const NodeInfo* node_info) {
+  uint64_t way_id = way_ids[way_id_index];
+
+  for (size_t j = 0; j < node_info->edge_count(); ++j) {
+    GraphId edge_id(tile->id().tileid(), tile->id().level(), node_info->edge_index() + j);
+    const DirectedEdge* de = tile->directededge(edge_id);
+
+    if (de->endnode() != prev_node &&
+        !(de->IsTransitLine() || de->is_shortcut() || de->use() == Use::kTransitConnection ||
+          de->use() == Use::kEgressConnection || de->use() == Use::kPlatformConnection)) {
+      auto edge_info = tile->edgeinfo(de->edgeinfo_offset());
+      if (edge_info.wayid() == way_id) {
+        edge_ids.push_back({way_id, edge_id});
+        bool found;
+
+        // expand with the next way_id
+        found = ExpandFromNode(reader, lock, last_node, edge_ids, way_ids, way_id_index + 1, tile,
```

Comment:
Calling `ExpandFromNode(..., way_id_index + 1, ...)` before `ExpandFromNode(..., way_id_index, ...)` let us not to worry about multiple edges with the last way id. So we need not remove duplicated way_ids in the prefix only